### PR TITLE
Feature/ministryplatformapi

### PR DESCRIPTION
**Pull request by silbermm**

### CODE REVIEW

File: lib/ministry_platform/ministry_platform_api.ex [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+      {:ok, %HTTPoison.Response{"body": body, "status_code": 200}} ->
+        email_addresses = body |> Poison.Parser.parse! |> Enum.map(&(&1["Email_Address"]))
+        {:ok, email_addresses}
+      {:ok, %HTTPoison.Response{"body": body, "status_code": 500}} ->
+        {_,_} = Logger.info("User does not have access to table, or table does not exist")
+        {:error, []}
+      {:error, _} ->
+        {:error, []}
+    end
+
```
Comment (Ganonside) [inline lib/ministry_platform/ministry_platform_api.ex]: inconsistent whitespace

File: lib/ministry_platform/ministry_platform_api.ex [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+  end
+
+  def get_auth_body do
+    ["username=#{URI.encode(@username)}",
+     "&password=#{URI.encode(@password)}",
+     "&client_id=#{@client_id}",
+     "&client_secret=#{@client_secret}",
+     "&grant_type=password"] |> Enum.join
+  end
+end
```
Comment (Ganonside) [inline lib/ministry_platform/ministry_platform_api.ex]: consider making some methods private

File: lib/tollbooth.ex [hunk 3]
```diff
@@ -1,4 +1,11 @@
 defmodule TollBooth do
+  @moduledoc """
+    Tollbooth is a specialized application that sits in between MinistryPlatform and Sendgrid.
+    
+    It handles all SMTP traffic from Ministry Platform (MP), checks a whitelist of allowed recipients and forwards that allowed emails to Sendgrid. All emails to unallowed recipients show in MP as sent, but never gets sent to Sendgrid.
```
Comment (Ganonside) [inline lib/tollbooth.ex]: consider wrapping this line to improve readability

File: lib/tollbooth_mail_parser.ex [hunk 4]
```diff
@@ -43,57 +45,51 @@ defmodule TollBooth.MailParser do
   def handle_cc(email, mail_map, allowed_list) do
     case Map.get(mail_map, "Cc") do
       nil -> email
-      _ = ccs -> 
-        cc_list =  String.split(ccs, ",") 
-            |> Enum.map(fn(x) -> String.strip(x) end)
-            |> Enum.filter(fn(x) -> Enum.member?(allowed_list,x) end)
-        email |> Email.add_cc(cc_list)
+      _ = ccs ->
```
Comment (Ganonside) [inline lib/tollbooth_mail_parser.ex]: why is the '_' here?

File: lib/tollbooth_mail_parser.ex [hunk 5]
```diff
@@ -43,57 +45,51 @@ defmodule TollBooth.MailParser do
...
-
+    
     email = email
       |> Email.put_to(List.first(recipients))
       |> Email.put_subject(build_subject(mail_map))
-      |> Email.put_from(Map.get(mail_map, "From") |> Map.get(:email))
-      |> Email.put_from_name(Map.get(mail_map, "From") |> Map.get(:name))
+      |> Email.put_from(mail_map |> Map.get( "From") |> Map.get(:email))
+      |> Email.put_from_name(mail_map |> Map.get("From") |> Map.get(:name))
       |> Email.put_reply_to(Map.get(mail_map, "Reply-To", email.from))
```
Comment (Ganonside) [inline lib/tollbooth_mail_parser.ex]: Inconsistent styling of pipes on new lines. Compare to lines 50 and 70

File: lib/tollbooth_smtp_server.ex [hunk 6]
```diff
@@ -14,13 +18,13 @@ defmodule TollBooth.Smtp.Server do
   @smtp_requested_action_okay 250
   @smtp_mail_action_abort 552
   @smtp_unrecognized_command 500
-  @env if (Application.get_env(:tollbooth, :environment) != nil), do:
-    Application.get_env(:tollbooth, :environment) |> String.downcase, else: "local"
-  @available_envs Application.get_env(:tollbooth, :available_environments, [])
+
+  @env if Application.get_env(:tollbooth, :environment) != nil, do:
+    :tollbooth |> Application.get_env(:environment) |> String.downcase, else: "local"
+  @available_envs :tollbooth |> Application.get_env(:available_environments)
```
Comment (Ganonside) [inline lib/tollbooth_smtp_server.ex]: should pipes be inline or on new lines?

File: lib/tollbooth_smtp_server.ex [hunk 7]
```diff
@@ -30,43 +34,45 @@ defmodule TollBooth.Smtp.Server do
...
-      end
-
-      if Enum.count(to) > 0 do
-        Logger.debug "#{to} should get an email"
-        email = parse_mail(data, state, unique_id)
-          |> MailParser.convert_to_email(state.recipients, allowed_list)
-        Mailer.send(email)
-      end
-      {:ok, unique_id, %{state | recipients: []}}
+          {:ok, unique_id, %{state | recipients: []}}
```
Comment (Ganonside) [inline lib/tollbooth_smtp_server.ex]: maybe use a new line to make this look cleaner?

File: lib/tollbooth_smtp_server.ex [hunk 8]
```diff
@@ -140,4 +134,5 @@ defmodule TollBooth.Smtp.Server do
         :io.format("Message decode FAILED with ~p:~n", [reason])
     end
   end
+
```
Comment (Ganonside) [inline lib/tollbooth_smtp_server.ex]: inconsistent white space

File: config/config.exs [hunk 9]
```diff
@@ -1,19 +1,28 @@
...
+  api_password: System.get_env("MP_API_PASSWORD"),
+  api_client_id: System.get_env("MP_API_CLIENTID"),
+  api_client_secret: System.get_env("MP_API_CLIENTSECRET")
 
 config :tollbooth, Repo,
-  database: System.get_env("MPDB_DATABASE"),
-  username: System.get_env("MPDB_USERNAME"),
-  password: System.get_env("MPDB_PASSWORD"),
-  hostname: System.get_env("MPDB_HOSTNAME")
+  database: System.get_env("MP_DATABASE"),
```
Comment (k8tied1) [inline config/config.exs]: I dont't think that we need to have the database connection info anymore

File: lib/ministry_platform/ministry_platform_api.ex (hunk 2 above)
Comment (silbermm) [inline lib/ministry_platform/ministry_platform_api.ex]: No thanks. I want to keep them public for testing purposes

File: lib/tollbooth_mail_parser.ex (hunk 5 above)
Comment (silbermm) [inline lib/tollbooth_mail_parser.ex]: these pipes are parameters to Email.put_from

**Pull request merged**